tfhe/src/integer/gpu/list_compression/server_keys.rs

```diff
@@ -0,0 +1,328 @@
+                lwe_dimension,
+                lwe_pksk.decomposition_base_log(),
+                lwe_pksk.decomposition_level_count(),
+                self.lwe_per_glwe.0 as u32,
+                self.storage_log_modulus.0 as u32,
+                num_lwes as u32,
+            );
+        }
+
+        (output_glwe, info, self.storage_log_modulus)
+    }
+}
+
+impl CudaDecompressionKey {
+    pub fn unpack(
+        &self,
+        packed_list: &(
+            CudaGlweCiphertextList<u64>,
+            Vec<CudaBlockInfo>,
+            CiphertextModulusLog,
+        ),
+        start_block_index: usize,
+        end_block_index: usize,
+        streams: &CudaStreams,
+        gpu_index: u32,
+    ) -> CudaRadixCiphertext {
+        let indexes_array = (start_block_index..=end_block_index)
+            .map(|x| x as u32)
+            .collect_vec();
+
+        let encryption_glwe_dimension = self.parameters.glwe_dimension();
+        let encryption_polynomial_size = self.parameters.polynomial_size();
+        let compression_glwe_dimension = packed_list.0.glwe_dimension();
+        let compression_polynomial_size = packed_list.0.polynomial_size();
+        let lwe_ciphertext_count = LweCiphertextCount(end_block_index - start_block_index + 1);
+        let message_modulus = self.parameters.message_modulus();
+        let carry_modulus = self.parameters.carry_modulus();
+        let ciphertext_modulus = self.parameters.ciphertext_modulus();
+        let storage_log_modulus = packed_list.2;
+
+        match &self.blind_rotate_key {
+            CudaBootstrappingKey::Classic(bsk) => {
+                let lwe_dimension = bsk.output_lwe_dimension();
+
+                let mut output_lwe = CudaLweCiphertextList::new(
+                    lwe_dimension,
+                    lwe_ciphertext_count,
+                    ciphertext_modulus,
+                    streams,
+                );
+
+                unsafe {
+                    let d_indexes_array =
+                        CudaVec::from_cpu_async(indexes_array.as_slice(), streams, gpu_index);
+
+                    compression_decompress_integer_radix_async(
+                        streams,
+                        &mut output_lwe.0.d_vec,
+                        &packed_list.0 .0.d_vec,
+                        &bsk.d_vec,
+                        message_modulus,
+                        carry_modulus,
+                        encryption_glwe_dimension,
+                        encryption_polynomial_size,
+                        compression_glwe_dimension,
+                        compression_polynomial_size,
+                        lwe_dimension,
+                        bsk.decomp_base_log(),
+                        bsk.decomp_level_count(),
+                        self.lwe_per_glwe.0 as u32,
+                        storage_log_modulus.0 as u32,
+                        &d_indexes_array,
+                        lwe_ciphertext_count.0 as u32,
+                    );
+                }
```

Comment:
Synchronization needed, or the function should be unsafe and async.

Reply:
I added a synchronization call before returning. I think this method does not need to be async.